Add millisecond option to `ConfigDict.ser_json_timedelta`

ollz272 opened the pull request:
<!-- Thank you for your contribution! -->
<!-- Unless your change is trivial, please create an issue to discuss the change before creating a PR -->

## Change Summary

Implements https://github.com/pydantic/pydantic/issues/10256 : adds a millisecond option to `ser_json_timedelta`. These are the changes needed in pydantic to enable the changes: see https://github.com/pydantic/pydantic-core/pull/1427 for the corrosponding pydantic core changes.

## Related issue number

fix #10256

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**


Selected Reviewer: @sydney-runkle

ollz272 commented:
Note that tests are not *yet* passing, but i've tested the new tests locally against the pydantic core changes and all looks good there. Just needs https://github.com/pydantic/pydantic-core/pull/1427 looked at and merged so this can get in.

sydney-runkle commented on pydantic/config.py:
```diff
@@ -563,7 +563,7 @@ class Transaction(BaseModel):
-    ser_json_timedelta: Literal['iso8601', 'float']
+    ser_json_timedelta: Literal['iso8601', 'float', 'millisecond']
     """
     The format of JSON serialized timedeltas. Accepts the string values of `'iso8601'` and
```
Will need to change the docstring here.

I'm wondering if we should add support for `'second'` as well, and deprecate `'float'` down the line (equivalent to `'second'`...

sydney-runkle reviewed (changes requested):
Looks good overall, though let's please:
* Update the docs
* Consider adding a `'second'` option, equivalent to `'float'` as mentioned. We could even deprecate `'float'` in v2 with `'second'` as the replacement.

Thanks for your contribution here!!

sydney-runkle commented:
Maybe `'seconds'` and `'milliseconds'` makes more sense...

sydney-runkle commented:
Will get back to you re this API change tomorrow morning 👍 - need to get feedback from the team re official decision.

ollz272 commented:
> Will get back to you re this API change tomorrow morning 👍 - need to get feedback from the team re official decision.

Great thank you, changes on the pydantic core side should be straightforward if you wanna do this. I'll update the docs

sydney-runkle commented:
Verdict on the API - let's go with:

`iso8601`
`seconds_float`
`milliseconds_float`

For now. We'll need to add deprecation info / docs updates for the `float -> seconds_float` logic. This also provides us with the ability to add `seconds_int` and `milliseconds_int` down the line if that's a high enough demand feature request.

Thanks for your help on this - happy to do another review once you've made those changes!

ollz272 commented:
> Verdict on the API - let's go with:
> 
> `iso8601` `seconds_float` `milliseconds_float`
> 
> For now. We'll need to add deprecation info / docs updates for the `float -> seconds_float` logic. This also provides us with the ability to add `seconds_int` and `milliseconds_int` down the line if that's a high enough demand feature request.
> 
> Thanks for your help on this - happy to do another review once you've made those changes!

Would you like me to just add `seconds_int` and `milliseconds_int`? Wouldn't be too much extra work 😄

ollz272 reviewed (commented):
Left some of my own comments explaining decisions.

ollz272 commented:
@sydney-runkle Hi!

Thanks for the response. I've now:

- Pinned this to my branch of pydantic core so tests can run in the pipeline properly
- Added seconds_float and milliseconds_float options
- Deprecated float, with docs and a warning
- Updated tests

Let me know if theres anything further you want me to do. And if you want me to add the int options, more than happy to just get this done 👍🏻 

Assigning back to you

ollz272 commented:
please review

sydney-runkle commented on pydantic/config.py:
```diff
@@ -563,13 +563,16 @@ class Transaction(BaseModel):
+    - `'float_seconds'` will serialize timedeltas to the total number of seconds.
+    - `'float_milliseconds'` will serialize timedeltas to the total number of seconds.
     - `'float'` will serialize timedeltas to the total number of seconds.
+        NOTE: `'float' is deprecated in favour of `'float_seconds'`
```
Let's list deprecated in v2.10

sydney-runkle reviewed (changes requested):
Nice work here! A couple of things:

* Let's fully deprecate `float` support - see comments above.
* Make sure that you're consistent with `seconds_float` vs `float_seconds`. I'd opt for the former.

DH had some good feedback on the `pydantic-core` end of things re multiplication with floats + performance.

Happy to do one more review here after you make the remaining changes.

davidhewitt commented on pydantic/_internal/_config.py:
```diff
@@ -169,6 +169,13 @@ def core_config(self, obj: Any) -> core_schema.CoreConfig:
 
+        if config.get('ser_json_timedelta') == 'float':
+            warnings.warn(
+                'Using a `float` as an option for `ser_json_timedelta` is deprecated, use a `float_seconds` instead.',
```
We should continue to emit this warning, even if we remove "support". We can remove the coercion in v3, I think.

ollz272 commented:
Right, i've pushed changes related to comments made

I think probably not worth reviewing again until the `pydantic-core` changes are finished.

Only thing to note is there are some type checking errors happening, though i think they're caused by https://github.com/pydantic/pydantic-core/pull/1443

sydney-runkle commented:
@ollz272,

Sounds good, ping me when you're ready for a final review here.

Cc @Viicos re:

> Only thing to note is there are some type checking errors happening, though i think they're caused by https://github.com/pydantic/pydantic-core/pull/1443

sydney-runkle commented on pyproject.toml:
```diff
@@ -50,9 +50,7 @@ dependencies = [
-    "pydantic-core==2.23.1",
-    # See: https://docs.python.org/3/library/zoneinfo.html#data-sources
-    'tzdata; python_version >= "3.9"',
+    "pydantic-core @ git+https://github.com/ollz272/pydantic-core@add-millisecond",
```
@ollz272, could you revert this change as well as the `pdm.lock` changes? Thanks!

sydney-runkle commented on pydantic/config.py:
```diff
@@ -562,13 +562,15 @@ class Transaction(BaseModel):
     """
 
-    ser_json_timedelta: Literal['iso8601', 'float']
+    ser_json_timedelta: Literal['iso8601', 'milliseconds_float', 'seconds_float']
```
```suggestion
    ser_json_timedelta: Literal['iso8601', 'seconds_float', 'milliseconds_float']
```

Nit pick, let's keep the order the same as that of `pydantic-core`

sydney-runkle commented on pydantic/config.py:
```diff
@@ -562,13 +562,15 @@ class Transaction(BaseModel):
     """
     The format of JSON serialized timedeltas. Accepts the string values of `'iso8601'` and
-    `'float'`. Defaults to `'iso8601'`.
+    'seconds_float'`, `'float_milliseconds'` and `'float'`. Defaults to `'iso8601'`.
```
Should read:

```
The format of JSON serialized timedeltas. Accepts the string values of `'iso8601'`, `'seconds_float'`, and `'milliseconds_float'`. Defaults to `'iso8601'`.
```

sydney-runkle commented on tests/test_json_schema.py:
```diff
@@ -1778,21 +1779,42 @@ class Outer(BaseModel):
-def test_model_default_timedelta(ser_json_timedelta: Literal['float', 'iso8601'], properties: typing.Dict[str, Any]):
-    class Model(BaseModel):
-        model_config = ConfigDict(ser_json_timedelta=ser_json_timedelta)
+def test_model_default_timedelta(
```
Yeah, I think having their own tests would be good so that we can avoid the conditional warning checks.

sydney-runkle reviewed (changes requested):
Looking great - just a few minor change requests, otherwise, we should be ready to merge!

I have some changes that I'd like to make to `pydantic-core` in the next week, so might wait for that to do the minor release with your changes.

ollz272 commented:
> Looking great - just a few minor change requests, otherwise, we should be ready to merge!
> 
> I have some changes that I'd like to make to `pydantic-core` in the next week, so might wait for that to do the minor release with your changes.

Makes sense, i think the only reason the docs ci is failing is because its not on a release of pydantic core. I'll fix up those points now. Happy to then change the version of pydantic core this points at, or for you guys to do it when you're ready for the pydantic core release.

sydney-runkle commented:
> Happy to then change the version of pydantic core this points at, or for you guys to do it when you're ready for the pydantic core release.

We can do this! If you could revert the `pdm.lock` and version changes though, that'd be great. Thanks!

ollz272 commented:
@sydney-runkle Hi!

Thanks, i've made your requested changes. Tests were still passing when i checked.

I've reverted the pyproject.toml and pdm.lock changes, so should be a simple change when you're ready.

Anything else let me know!

sydney-runkle reviewed (approved):
Amazing work here. Looking forward to this change.

I'll try to get a `pydantic-core` release out this week so that we can push this across the line.

sydney-runkle commented:
cc @ollz272, all set to merge :)
